virtual_ecosystem/models/litter/input_partition.py

```diff
@@ -1,32 +1,197 @@
+
+@dataclass
+class InputDetails:
+    """The full details of the input to the litter model."""
+
+    leaf_mass: NDArray[np.float32]
+    """Total leaf input mass to litter [kg C m^-2]"""
+    root_mass: NDArray[np.float32]
+    """Total root input mass to litter [kg C m^-2]"""
+    deadwood_mass: NDArray[np.float32]
+    """Total deadwood input mass to litter [kg C m^-2]"""
+    reprod_mass: NDArray[np.float32]
+    """Total plant reproductive tissue input mass to litter [kg C m^-2]"""
+
+    leaf_lignin: NDArray[np.float32]
+    """Lignin proportion of leaf input [unitless]"""
+    root_lignin: NDArray[np.float32]
+    """Lignin proportion of root input [unitless]"""
+    deadwood_lignin: NDArray[np.float32]
+    """Lignin proportion of deadwood input [unitless]"""
+    reprod_lignin: NDArray[np.float32]
+    """Lignin proportion of reproductive tissue input [unitless]"""
+
+    leaf_nitrogen: NDArray[np.float32]
+    """Carbon nitrogen ratio of leaf input [unitless]"""
+    root_nitrogen: NDArray[np.float32]
+    """Carbon nitrogen ratio of root input [unitless]"""
+    deadwood_nitrogen: NDArray[np.float32]
+    """Carbon nitrogen ratio of deadwood input [unitless]"""
+    reprod_nitrogen: NDArray[np.float32]
+    """Carbon nitrogen ratio of reproductive tissue input [unitless]"""
+
+    leaf_phosphorus: NDArray[np.float32]
+    """Carbon phosphorus ratio of leaf input [unitless]"""
+    root_phosphorus: NDArray[np.float32]
+    """Carbon phosphorus ratio of root input [unitless]"""
+    deadwood_phosphorus: NDArray[np.float32]
+    """Carbon phosphorus ratio of deadwood input [unitless]"""
+    reprod_phosphorus: NDArray[np.float32]
+    """Carbon phosphorus ratio of reproductive tissue input [unitless]"""
+
+    leaves_meta_split: NDArray[np.float32]
+    """Fraction of leaf input that goes to metabolic litter [unitless]"""
+    reproduct_meta_split: NDArray[np.float32]
+    """Fraction of leaf input that goes to metabolic litter [unitless]"""
+    roots_meta_split: NDArray[np.float32]
+    """Fraction of leaf input that goes to metabolic litter [unitless]"""
+
+    input_woody: NDArray[np.float32]
+    """Total input to the woody litter pool [kg C m^-2]"""
+    input_above_metabolic: NDArray[np.float32]
+    """Total input to the above ground metabolic litter pool [kg C m^-2]"""
+    input_above_structural: NDArray[np.float32]
+    """Total input to the above ground structural litter pool [kg C m^-2]"""
+    input_below_metabolic: NDArray[np.float32]
+    """Total input to the below ground metabolic litter pool [kg C m^-2]"""
+    input_below_structural: NDArray[np.float32]
+    """Total input to the below ground structural litter pool [kg C m^-2]"""
+
+
+class InputPartition:
+    """This class handles the partitioning of plant matter between litter pools.
+
+    This class contains methods to calculate the partition of input plant matter between
+    the different litter pools based on the contents of the `data` object.
+    """
```

Comment:
This feels structurally a bit odd - the `InputPartition` class really only exists as a factory method for `InputDetails` and as a thin wrapper around `Data`. I can think of two alternative structures:

* Make `Data` and `constants` as `InitVar` attributes to `InputDetails` and make all the rest of the attributes `field(init=False)`. Then all of the logic in `InputPartition` can be put into a `__post_init__` method for `InputDetails`.

* That does restrict you to only ever populating an instance of `InputDetails` using `Data`, which to be honest sounds like it is the only use case. But you could also have it as a factory class method on the current dataclass. You could then create an instance programatically from dicts of variables.

```python3
@dataclass
class InputDetails:
    ...
    
    @classmethod
    def create_from_data(cls, data: Data, constants: LitterConsts) -> InputDetails:
         ...
```
  You would also need to pass `Data` to `combine_input_sources` - that could then just be a standalone function.